### PR TITLE
DBManager Refactor and simplification of ExoMol MdBMol

### PR DESCRIPTION
**Pull request by erwanp**
### List of changes : 
- RADIS's [MdBExomol](https://radis.readthedocs.io/en/latest/source/radis.io.exomol.html#radis.io.exomol.MdbExomol) was Refactored 
- more abstraction in DBManager load conditions
- HDF5Manager generalized to DataFileManager (compatbile with Feather)
- MdBExoMol uses DbManager  (no more Vaex/Feather/Pytables code duplicate)

# Discussion 

@HajimeKawahara

Using this PR to introduce a few ideas in preparation of the RADIS/ExoJax meeting where we'll talk about having a common API for database management. 

### Rationale

I'd like to keep 3 layers of agnosticism : 

Users shall not have to worry about : 
1. the **format of the input database** (HITRAN 2012 / HITRAN 2020 / / HITEMP / GEISA / ExoMol / Kuruz/ NIST / etc)
	(so that new librairies can be added easily; so that all codes benefit from all librairies)
2. the **format stored on disk** (Vaex's HDF5, Pandas's HDF5, Feather, etc.)
	(to be able to switch to more performant librairies when available. PyArrow etc.)  
3. the **output format used in calculations** (Pandas's DataFrame, Jax arrays, Vaex DataFrame)
	(to be able to retrieve data in the format needed by each code, i.e. `jax` (ExoJax) ; `pandas` (current RADIS), `vaex` (future fully-out-of-core RADIS for extreme databases) )

Current status for all of these : 

1.	ExoJax has [MdbExomol](https://github.com/HajimeKawahara/exojax/blob/master/src/exojax/spec/moldb.py#L18) or MdbHIt . They are similar but don’t inherit from a common base ; 
RADIS 0.12 has a [DatabaseManager](https://radis.readthedocs.io/en/latest/source/radis.io.dbmanager.html#radis.io.dbmanager.DatabaseManager) to handle HITEMP 2020 / HITRAN 2020; the ExoMol support is done with a MdBMol taken from Exojax with no link with the DatabaseManager. 
2.	RADIS implemented a DataFileManager class to handle (Vaex’s HDF5; Pandas’s HDF5, Feather) ;   DatabaseManager calls DataFileManager  internally ✔️ 
3.	Radis #464 added 3 output : ’jax/vaex/pandas' to RADIS’s DatabaseManager

Thesis : 
- [DatabaseManager](https://radis.readthedocs.io/en/latest/source/radis.io.dbmanager.html#radis.io.dbmanager.DatabaseManager) could be the class needed to handle 1. all input databases; handling internally 2. all formats; and giving back 3. any format . 

**Building a common, generalized architecture is a very long process. We would get stuck by overthinking it from the start. Let's do it by trying, and fix problems as they arise along the way.** 

This PR is a first step in that direction. It is a proof-of-concept. Is it designed to be criticized, and commented. @HajimeKawahara @minouHub @ykawashima or anyone else, feel free to fire your comments and ideas ! 

### Current Implementation details : 

RADIS's [MdBExomol](https://radis.readthedocs.io/en/latest/source/radis.io.exomol.html#radis.io.exomol.MdbExomol) was totally refactored and updated, to use the RADIS DatabaseManager. This is done on purpose: MdBExomol was originally taken from Exojax, and is updated here to show how Exojax's underlying API could evolve. 

It introduces a few additions which I personnaly think are good :)  :  see also the various comments in the code below, and comment there : https://github.com/radis/radis/pull/465/files#diff-0ccf9b5bb7915e8940d04891456118106d4cc97e3052e8d77fc2b7b85673ba6b   (the GitHub review page is great to understand the differences & suggestions)

**Use a DataFrame directly (rather than arrays).** 

- Makes Masking implementation easier and N times faster (where N is the number of arrays ) 
- uses less memory. 
- Scales well for very large databases 
- Easier to add new parameters / columns (no need to worry about index)

It is a proof of concept. 

It already allows to : 

```python
# Test ExoMol

import radis
radis.config["MEMORY_MAPPING_ENGINE"] = "vaex" # 👉👉👉  choose  "vaex", "pytables", "feather"

from radis.io.exomol import fetch_exomol

df = fetch_exomol("SiO", database="EBJT", isotope="1", load_wavenum_max=5000, 
        output="pandas" # 👉👉👉 choose : "pandas", "vaex", "jax"
        )
print(df)
```

```python
# Test HITEMP

import radis
radis.config["MEMORY_MAPPING_ENGINE"] = "vaex" # 👉👉👉  choose  "vaex", "pytables", "feather"

from radis.io.hitemp import fetch_hitemp

df = fetch_hitemp("OH", 
        output="pandas" # 👉👉👉 choose : "pandas", "vaex", "jax"
        )
print(df)
```

### In the future : 

- RADIS's DBManager could easily be called by ExoJax).  RADIS has low requirements, can be used in any OS 
- In any case, ExoJax should subclass RADIS’s DataFileManager  and DatabaseManager; just in case some changes should be done in ExoJax, without having to wait for a new RADIS release 
- will also help to have both codes RADIS & Exojax run on the same databases, with the same functions --> helps validation  (I actually have a `calc_spectrum_exojax()` implemented (not here), mirroring RADIS's calc_spectrum, to compare outputs. 
-Helps make new databases  (GEISA, HITRAN 2020)  available to all codes 
-May also help to deploy ExoJax on RADIS preconfigured database systems (RADIS-Lab – although no GPU yet -  or future Co-Lab versions )
- Exojax could also make sure of RADIS's config file system, with user preferences (such as : local_database folder, favorite plotting library ; default local file format )

### Current questions 

- Is this the right direction ? 
- Would it help Exojax too ? 
- How hard would it be to convert some of the current ExoJax MdB to this ? 

---

# EDIT : 

Answer & Next : 
- https://github.com/HajimeKawahara/exojax/issues/256 
- https://github.com/HajimeKawahara/exojax/discussions/257

### CODE REVIEW

File: radis/io/exomol.py [hunk 1]
```diff
@@ -558,32 +654,22 @@ def __init__(
...
-            elif self.states_file.with_suffix(".hdf5").exists():
-                states_file_cache = self.states_file.with_suffix(".hdf5")
-                states = vaex.open(states_file_cache)
-            else:
-                print(
-                    "Note: Caching states data to the hdf5 format with vaex. After the second time, it will become much faster."
-                )
-                states = exomolapi.read_states(self.states_file, dic_def, engine="vaex")
-            ndstates = vaex.array_types.to_numpy(states)
+        if mgr.cache_file(self.states_file).exists():
```
Comment (erwanp) [inline radis/io/exomol.py:660]: only one syntax for any on-file format (HDF5, Feather, etc)

File: radis/io/exomol.py [hunk 2]
```diff
@@ -719,191 +701,57 @@ def __init__(
...
-                        #    )  # merge without column name will merge on rows basis
-
-                        # mask needs to be applied
-                        mask_needed = True
-
-                    ndtrans = vaex.array_types.to_numpy(trans)
                 self.trans_file.append(trans_file)
+                self.num_tag.append(dic_def["numtag"][i])
+
+        # Look-up missing parameters and write file
```
Comment (erwanp) [inline radis/io/exomol.py:712]: Separate Find Filenames (1st) and Download & write them (after) --> helps remove copies of code, more readable

File: radis/io/exomol.py [hunk 3]
```diff
@@ -719,191 +701,57 @@ def __init__(
...
-                        self.Sij0 = linestrength_from_Einstein(
-                            A=self._A,
-                            gu=self._gpp,
-                            El=self._elower,
-                            Ia=1,  #  Sij0 is a linestrength calculated without taking into account isotopic abundance (unlike line intensity parameter of HITRAN. In RADIS this is corrected for in fetch_exomol()  )
-                            nu=self.nu_lines,
-                            Q=self.QTref,
-                            T=self.Tref,
-                        )
+                exomolapi.pickup_gE(states, trans, trans_file, dic_def)
```
Comment (erwanp) [inline radis/io/exomol.py:732]: DataFrame is updaated --> no need to build and maintain all arrays

File: radis/io/hdf5.py [hunk 4]
```diff
@@ -209,6 +218,94 @@ def __del__(self):
...
+            if columns:  # load only these columns (if they exist)
+                columns = [c for c in columns if c in df.columns]
+            if output == "pandas":
+                df = df.to_pandas_df(column_names=columns)
+            elif output == "jax":
+                if columns == None:
+                    columns = ["wav", "int", "A", "El", "gpp"]
+                out = {}
+                try:
+                    import jax as jnp
```
Comment (HajimeKawahara) [inline radis/io/hdf5.py:284]: This should be `import jax.numpy as jnp`.

Comment (HajimeKawahara) [reply radis/io/hdf5.py:284]: @erwanp Can you fix this part? Then, I will try to make a prototype of the conversion from a DataFrame to ExoJAX.

Comment (erwanp) [reply radis/io/hdf5.py:284]: Fixed !

**Comment by erwanp** [comment]
btw @mariottop you can enforce code linting automatically by installing `pre-commit` for Radis : https://radis.readthedocs.io/en/latest/dev/developer.html?highlight=linting#code-linting   .  Will fix all the Linting tests.

**Comment by erwanp** [comment]
Merging to `develop` as the changes contain important ExoMol fixes. 

The discussion on Common API continue in https://github.com/HajimeKawahara/exojax/discussions/257
